### PR TITLE
Expose GpuElemwise from python and remove ElemwiseKernel

### PR DESCRIPTION
**Pull request by abergeron**
Since we now have the C version for elemwise, we can get rid of the experimental python version.

This should also be appreciatively faster.

### CODE REVIEW

File: src/gpuarray_elemwise.c [hunk 1]
```diff
@@ -561,57 +562,55 @@ GpuElemwise *GpuElemwise_new(gpucontext *ctx,
...
+  res->nd = 8;
+  res->n = n;
 
   res->expr = strdup(expr);
   if (res->expr == NULL)
-    goto fail_expr;
-  if (preamble == NULL) {
-    res->preamble = NULL;
-  } else {
+    goto fail;
```
Comment (nouiz) [inline src/gpuarray_elemwise.c]: To be safe, on line 561 you should call calloc, as otherwise res isn't zeroed and if there is an error, the new free code can have problems.

File: pygpu/gpuarray.pyx [hunk 2]
```diff
@@ -1942,9 +1942,9 @@ cdef class GpuKernel:
                     _types[i] = GA_BUFFER
                 else:
                     _types[i] = dtype_to_typecode(types[i])
-                self.callbuf[i] = malloc(gpuarray_get_elsize(_types[i]))
-                if self.callbuf[i] == NULL:
-                    raise MemoryError
+                    self.callbuf[i] = malloc(gpuarray_get_elsize(_types[i]))
```
Comment (nouiz) [inline pygpu/gpuarray.pyx]: On line 1891, you don't free the those malloc. Just callbuf itself.

File: pygpu/tests/test_elemwise.py [hunk 3]
```diff
@@ -101,87 +100,6 @@ def ielemwise2_ops_array(op, dtype1, dtype2, shape):
     assert numpy.allclose(out_c, numpy.asarray(out_g), atol=1e-6)
 
 
-def test_elemwise_layouts():
```
Comment (nouiz) [inline pygpu/tests/test_elemwise.py]: Why remove those tests? I think this tests more cases then the C tests. So I would have a tendency to keep them.

File: pygpu/tests/test_elemwise.py [hunk 4]
```diff
@@ -101,87 +100,6 @@ def ielemwise2_ops_array(op, dtype1, dtype2, shape):
...
-                for sliced in [1, 2]:
-                    for order in ['c', 'f']:
-                        yield elemwise_layouts, shape, offseted_outer, \
-                            offseted_inner, sliced, order
-                        yield elemwise_layouts_mixed, shape, offseted_outer, \
-                            offseted_inner, sliced, order
-
-
-def test_elemwise_0():
-    elemwise_layouts((0,), False, False, 1, 'c')
```
Comment (nouiz) [inline pygpu/tests/test_elemwise.py]: I don't think this is in the c test too.

File: pygpu/tests/test_elemwise.py [hunk 5]
```diff
@@ -328,117 +246,3 @@ def broadcast(shapea, shapeb):
     rg = ag + bg
 
     check_meta_content(rg, rc)
-
-
-def test_elemwise_collapse():
```
Comment (nouiz) [inline pygpu/tests/test_elemwise.py]: good idea to remove.

File: pygpu/tests/test_elemwise.py [hunk 6]
```diff
@@ -7,9 +7,9 @@
 from six import PY2
 
 from .support import (guard_devsup, rand, check_flags, check_meta, check_all,
-                      context, gen_gpuarray, dtypes_no_complex,
-                      check_meta_content)
+                      context, gen_gpuarray, check_meta_content)
 
+dtypes_test = ['float32', 'int32', 'uint64']
```
Comment (nouiz) [inline pygpu/tests/test_elemwise.py]: I woud also test int8. This is corner case that already caused problems.

File: src/gpuarray_elemwise.c (hunk 1 above)
Comment (abergeron) [inline src/gpuarray_elemwise.c]: Yes

File: pygpu/tests/test_elemwise.py (hunk 3 above)
Comment (abergeron) [inline pygpu/tests/test_elemwise.py]: I would rather extend the C tests than keep those.

File: pygpu/tests/test_elemwise.py (hunk 6 above)
Comment (abergeron) [inline pygpu/tests/test_elemwise.py]: Ok.

File: pygpu/gpuarray.pyx (hunk 2 above)
Comment (abergeron) [inline pygpu/gpuarray.pyx]: I know.  It was like that before, but I'll fix it.

File: pygpu/tests/test_elemwise.py (hunk 3 above)
Comment (nouiz) [inline pygpu/tests/test_elemwise.py]: I'm fine with that.

File: pygpu/tests/test_elemwise.py (hunk 4 above)
Comment (abergeron) [inline pygpu/tests/test_elemwise.py]: No it's not.  I'm not sure I want to add it, because I'm not sure I want to support it.  Currently it doesn't work.

File: pygpu/tests/test_elemwise.py (hunk 3 above)
Comment (abergeron) [inline pygpu/tests/test_elemwise.py]: Now that I look at it in detail, it's almost all tested, but with smaller arrays.  I'll add a test for an offset, but it's all that's missing.  It might look like more things are tested but they are irrelevant and where only there because I figure I might as well use all the options of gen_gpuarray().

Now that I've seen that it makes for extra-long tests, I tend to be more minimalist.

File: pygpu/tests/test_elemwise.py (hunk 4 above)
Comment (nouiz) [inline pygpu/tests/test_elemwise.py]: If we don't support it there, then we will need to support it at every places that call elemwise. I prefer to support it at only 1 place...

File: tests/check_elemwise.c [hunk 7]
```diff
@@ -131,6 +131,72 @@ START_TEST(test_basic_simple)
...
+
+  ge = GpuElemwise_new(ctx, "", "c = a + b", 3, args, 2, 0);
+
+  ck_assert_ptr_ne(ge, NULL);
+
+  rargs[0] = &a;
+  rargs[1] = &b;
+  rargs[2] = &c;
+
+  ga_assert_ok(GpuElemwise_call(ge, rargs, GE_NOCOLLAPSE));
```
Comment (nouiz) [inline tests/check_elemwise.c]: call it with collapsing too.

Otherwise the test seem good.

Comment (abergeron) [inline tests/check_elemwise.c]: The collapsing code is tested elsewhere and doesn't interact with the offset.  I prefer to keep things separate if possible.

File: pygpu/tests/test_elemwise.py (hunk 4 above)
Comment (abergeron) [inline pygpu/tests/test_elemwise.py]: Yes, I suppose it's better to have that in one place.

Comment (abergeron) [inline pygpu/tests/test_elemwise.py]: I added support.

**Comment by nouiz** [comment]
You need to bump the feature version number and check that number in the corresponding Theano PR.

**Comment by nouiz** [comment]
Don't forget to make a comment when you update a PR. This PR need a rebase. After that, we can merge.

**Comment by abergeron** [comment]
Rebase is done, all should be well.
